### ISSUE TITLE
`CronJob` perpetually out of sync after `apiVersion: batch/v1` promotion

### ISSUE THREAD

**Issue by booleanbetrayal** [argoproj/argo-cd#6876]
**Checklist**

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

We updated our Kubernetes clusters to 1.21 and shortly after, updated our `CronJob` Helm templates to toggle `apiVersion` from `batch/v1beta1` to `batch/v1`. Upon syncing those changes, our `CronJob` resources reflected immaterial diffs, including diffs on the human-readable container resources values, and they remain that way through subsequent syncs and ArgoCD deployment restarts. We have not seen similar problems when promoting our `Ingress` resources, and we use the same resource values conventions everywhere, so we're a bit boggled.

**To Reproduce**

0. _Promote an ArgoCD managed Kubernetes cluster from 1.20 -> 1.21 (???)_.
1. Promote `CronJob` resources to `apiVersion: batch/v1` from  `apiVersion: batch/v1beta1`.
2. Sync the parrent ArgoCD application and observe immaterial diffs that persist through subsequent re-syncs.

**Expected behavior**

`CronJob` resources sync and do not present immaterial diffs.

**Screenshots**

Example `CronJob`:

<img width="1410" alt="Screen Shot 2021-08-01 at 3 22 41 PM" src="https://user-images.githubusercontent.com/981825/127785989-7dd139c6-5be8-473e-a614-ac1bfa1dc496.png">

**Version**

```shell
argocd: v2.0.5+4c94d88
  BuildDate: 2021-07-22T20:52:31Z
  GitCommit: 4c94d886f56bcb2f9d5b3251fdc049c2d1354b88
  GitTreeState: clean
  GoVersion: go1.16
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

_Can provide sanitized logs at request but did not see anything immediately illuminating in the various ArgoCD pod logs._

**Comment by rbreeze** [comment]
A workaround would be to add these fields to `ignoreDifferences`. See https://argoproj.github.io/argo-cd/user-guide/diffing/

**Comment by booleanbetrayal** [comment]
Thanks @rbreeze . I had briefly considered going that route, and have done it for things that truly do change / rotate out from under out management, but I wouldn't want to ever ignore the container resource values, for example.

**Comment by booleanbetrayal** [comment]
@rbreeze / @alexmt - This appears to be fixed as of `2.1.0`!

**Comment by alexmt** [comment]
Thank you!

**Comment by zerunhu** [comment]
@alexmt  hi， I am using argocd version: v2.3.4+ac8b7df seems to have this problem again
![image](https://user-images.githubusercontent.com/44351924/197925820-966218ae-6b99-4980-9d10-4ea0573dbc38.png)
![image](https://user-images.githubusercontent.com/44351924/197925974-6a4e226b-e074-4bdf-8318-7c126164198f.png)

**Comment by zerunhu** [comment]
The object I am using is a CRD who actually references corev1.Container{}. In the LIVE MANIFEST page of ArgoCD ui,  i observed that the field workload.template.containers[0].envs[0].valueFrom.fieldRef.apiVersion is masked in the LIVE MANIFEST page. I used kubectl get workload -o yaml to get this field with this apiVersion so there is no DIFF with DESIRED MANIFEST page. but not blocking this field for my CRD, is it possible that the CRD is being missed when processing the relevant fields?
The same is true for cpu: '4' and cpu: 4000. This object is also a workload that we have developed and the underlying layer is also used in appsv1.Deployment

**Comment by zerunhu** [comment]
I am using a SideCarSet CRD, which defines a k8s native corev1.Containers object, and injects the container into the workload pod when it is created, so its fields are the same as the pod.template, including the env. 
I use the `kubectl get workload xx -o yaml` command to check any workload and find that the envs[0].valueFrom.fieldRef.apiVersion field is present . However, on the LIVE MANIFEST page of the ArgoCD ui, I observed that the same workload field envis[0].valueFrom.fieldRef.apiVersion was masked, so there is no diff with DESIRED MANIFEST page, which is the yaml configuration I set up
So I think it is ArgoCD that blocks this apiVersion by some specific when fetching the workload resource, the resource will not have the outofSync. This SidecarSet CRD does not mask this field hence the DIFF situation, I'm guessing the logic of the code is not judgment the CRD?
The same is true for cpu: '4' and cpu: 4000. This object is also a workload that we have developed and the underlying layer is also used in appsv1.Deployment
